### ISSUE TITLE
Privacy mode

### ISSUE THREAD

**Issue by akominch** [rroller/dahua#379]
Hi, I have a DH-SD-H4C-0400B camera, it has a privacy mode where its lens goes down. When I enable this mode in DMSS, the following entry appears in the camera logs 
```
Name: leLensMask
State: Start
Result: Success
```
Maybe someone knows if there is any way to manage this locally via cgi or rpc?

**Comment by Hukuma1** [comment]
Just switched from Amcrest to Dahua integration  and would also love the ability to flip on the Privacy Mode switch for some cameras that I have that support it (model IP4M-1041W if it matters). The official Amcrest integration has it, would it be possible to port it from there maybe?

**Comment by beckynet** [comment]
I've DH-H4C This camera also has a Pivacy mode.

**Comment by Hukuma1** [comment]
@zios0007 This is a longshot, and I apologize for being rude, but I saw you do a bunch of PRs to this repo. Any chance you could please add this? The code already exists in the official repo, perhaps we could merge it over to this repo? Again, sorry for rudely pinging you directly.

**Comment by zios0007** [comment]
@Hukuma1 not rude at all, and you were closer than you knew — the code you are
asking for has existed in this repo for eleven months.

**#516, "Add Privacy Mode Toggle" by @AdrianCzebiniak**, opened 2025-10-12. It
adds a Privacy Mode switch driven by `LeLensMask` over RPC2 — which matches the
`Name: leLensMask` line @akominch found in their camera log at the top of this
issue. It is 301 lines across 10 files with its own tests.

It never stalled on review. @rroller left one comment, the author fixed it two
days later, and then `main` moved underneath it until it went `CONFLICTING`.
That is the whole story: it needed a rebase, not a decision.

So I have done that — **#682** carries @AdrianCzebiniak's commit, authorship
intact, rebased onto current `main`. The three conflicts were all against
changes of mine and all additive. If it merges, this issue closes with it.

Two honest caveats:

- I have no camera with `LeLensMask`, so this is unverified on hardware — as
  #516 was. You have an IP4M-1041W and @beckynet has a DH-H4C; once it is
  released, a "the lens actually drops" from either of you is the confirmation
  that is missing.
- It is @rroller's call whether to take it, and I have flagged one thing in the
  PR for him to look at — the RPC2 session is opened and closed per poll, which
  cuts against a lot of recent work reducing logins.

@beckynet @GGontijo @akominch — tagging you since you all asked here.

The credit is @AdrianCzebiniak's. All I did was move it onto a branch that
merges.

**Comment by zios0007** [comment]
Released. **Privacy mode is in 0.10.2**, out a few hours after this thread woke
up — #682 merged and closed this issue with it, two years and one month after
@akominch opened it.

The credit is @AdrianCzebiniak's; the code was written last October and needed
only a rebase.

**It is still unverified on hardware.** I have no camera with `LeLensMask`, so
nobody has yet watched a lens actually drop from Home Assistant. @Hukuma1 you
have an IP4M-1041W, @beckynet a DH-H4C, @akominch a DH-SD-H4C-0400B — if one of
you updates to 0.10.2 and reports back either way, that closes the loop properly.

What to expect: a **Privacy Mode** switch on cameras that support it. It is
probed at startup over RPC2, so a camera without `LeLensMask` simply will not get
the entity. If the switch does not appear on a camera you know has the feature,
that is worth reopening this for — with `custom_components.dahua` at debug, the
line to look for is `Device supports privacy mode=`.

**Comment by Hukuma1** [comment]
Thank you so much for looking into this!

So one quick UI thing, I think Privacy Mode is sharing the same description string.

<img width="1125" height="179" alt="Image" src="https://github.com/user-attachments/assets/e3cb9788-b5d8-45ae-a4f9-a2f29733e8a8" />

```
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/websocket_api/decorators.py", line 26, in _handle_async_response
    await func(hass, connection, msg)
  File "/usr/src/homeassistant/homeassistant/components/websocket_api/commands.py", line 1212, in handle_execute_script
    script_result = await script_obj.async_run(
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^
        msg.get("variables"), context=context
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 1998, in async_run
    return await asyncio.shield(create_eager_task(run.async_run()))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 503, in async_run
    await self._async_step(log_exceptions=False)
  File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 575, in _async_step
    self._handle_exception(
    ~~~~~~~~~~~~~~~~~~~~~~^
        trace_element,
        ^^^^^^^^^^^^^^
    ...<2 lines>...
        self._log_exceptions or log_exceptions,
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 612, in _handle_exception
    raise exception
  File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 573, in _async_step
    await getattr(self, handler)()
  File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 1081, in _async_step_call_service
    response_data = await self._async_run_long_action(
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    ...<9 lines>...
    )
    ^
  File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 683, in _async_run_long_action
    return await long_task
           ^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/core.py", line 2961, in async_call
    response_data = await coro
                    ^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/core.py", line 3004, in _execute_service
    return await target(service_call)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/service.py", line 914, in entity_service_call
    response_data = await _async_handle_entity_calls(
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    ...<5 lines>...
    )
    ^
  File "/usr/src/homeassistant/homeassistant/helpers/service.py", line 806, in _async_handle_entity_calls
    raise result from None
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1711, in async_request_call
    return await coro
           ^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/service.py", line 782, in _with_context
    return await coro
           ^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/service.py", line 866, in _handle_single_entity_call
    result = await task
             ^^^^^^^^^^
  File "/config/custom_components/dahua/camera.py", line 433, in async_set_privacy_mode
    await self._coordinator.client.async_set_privacy_mode(enabled)
  File "/config/custom_components/dahua/client.py", line 1161, in async_set_privacy_mode
    await self._async_privacy_mode_rpc2(
        lambda rpc2: rpc2.async_set_privacy_mode(enabled), "privacy mode write"
    )
  File "/config/custom_components/dahua/client.py", line 1143, in _async_privacy_mode_rpc2
    return await action(rpc2)
           ^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/dahua/rpc2.py", line 284, in async_set_privacy_mode
    table = await self._async_get_privacy_mode_table()
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/dahua/rpc2.py", line 267, in _async_get_privacy_mode_table
    params = await self.get_config({"name": "LeLensMask"})
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/dahua/rpc2.py", line 225, in get_config
    response = await self.request(method="configManager.getConfig", params=params)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/dahua/rpc2.py", line 73, in request
    raise ConnectionError(
    ...<3 lines>...
    )
ConnectionError: Dahua RPC2 method configManager.getConfig returned result=false (code=285278249, message=Authority:check failure.)
```

I figured I didn't have permissions on the camera side so I just gave the user admin priveleges with all permissions enabled, but I still got that error. :/

**Comment by zios0007** [comment]
Both of those are useful. Taking them in the order that matters.

**`Authority:check failure` is probably not about permissions.** You did the
right thing giving the account full rights, and the error surviving that is the
tell. Dahua's RPC2 returns this code for a config table the device does not
have, not only for one you are not allowed to read — an unknown `name` and a
forbidden `name` come back the same way.

Which raises the question of whether your IP4M-1041W has `LeLensMask` at all.
That table is the **physical lens mask** — the motorised cover that drops in front
of the lens, as on @akominch's DH-SD-H4C-0400B, whose camera log at the top of
this issue shows `Name: leLensMask`. A camera with no moving cover has nothing to
put in it.

**The command that settles it**, straight to the camera, nothing to do with Home
Assistant:

```bash
curl -s --digest -u "USER:PASS" \
  'http://CAMERA-IP/cgi-bin/configManager.cgi?action=getConfig&name=LeLensMask'
```

- **A table comes back** — then your camera has the feature and CGI can read what
  RPC2 refuses. That is a real defect and a fixable one: the implementation is
  RPC2-only, and a CGI path would serve your camera.
- **`Error / Bad Request`, or an empty reply** — then this camera has no lens
  mask, and no amount of fixing this feature will give you one.

Also worth checking: **did a "Privacy Mode" switch entity actually appear** for
this camera? The switch is only created when the startup probe succeeds, while
the `dahua.set_privacy_mode` *service* exists regardless — which is why you got a
raw traceback rather than a missing entity. If there is no switch, the probe
already failed the same way, and that is consistent with the second case above.

**On the description strings** — you are right that they are too close, and they
are two genuinely different features:

- `dahua.set_privacy_masking` → `PrivacyMasking[0][n].Enable`, which blanks
  **regions of the image**, and is the one Amcrest cameras generally have.
- `dahua.set_privacy_mode` → `LeLensMask`, which **physically covers the lens**.

"Enables or disabled the cameras privacy masking" and "Enables or disables the
camera's lens privacy mask" do not convey that at all, and I would have picked
the wrong one too. I will send a PR making them say which is which.

And if your camera turns out to have no lens mask, `dahua.set_privacy_masking` is
very likely the feature you actually wanted from the Amcrest integration —
worth trying on its own.

**Comment by Hukuma1** [comment]
Ah that makes it much more clear. Yeah, my camera doesn't have a hardware physical cover. Enabling Privacy Mode on my lower-end camera simply forces the camera position (PTZ/preset?) to go all the way up to cover itself (so guests don't see a lens pointing at them) and cuts the feed.

I did not see any switches exposed on my end thinking it may have been a detection bug. I tried to call the service directly. But now it makes sense, my camera doesn't have the lens covering hardware the other people in this thread have. :/

<img width="1525" height="697" alt="Image" src="https://github.com/user-attachments/assets/184f5488-2d35-44ff-b1a5-2917ad2f355e" />

**Comment by zios0007** [comment]
That explains it completely, and your camera's behaviour is worth writing down
here because it is a third thing again.

- @akominch's DH-SD-H4C-0400B: a **motorised cover** drops over the lens.
  `LeLensMask`, which is what #682 implements.
- Yours: the camera **points itself away** and cuts the feed. No mask table, no
  cover — a PTZ move plus a video stop, dressed up as one button in the app.

So no switch appeared because the startup probe asked for `LeLensMask`, your
camera said it has no such thing, and the entity was correctly not created. The
service still existed, which is why calling it gave you a traceback rather than
nothing — that asymmetry is real and worth tidying, but it is not what was
stopping you.

**You can very likely rebuild your camera's version yourself today.** What the
app does is two actions this integration already exposes:

1. Move to a preset pointing at the ceiling — the **Preset Position** select
   entity, if your camera reports presets. Set one up in the camera's own web UI
   first, pointed where you want "privacy" to look.
2. Stop the picture — `dahua.set_privacy_masking` with a mask covering the whole
   frame, configured once under Privacy Masking in the camera's web interface.

Wrap the two in a script and you have a Privacy Mode button that matches what
DMSS does on your hardware. It is not as tidy as a single switch, and if enough
IP4M-class cameras turn out to work this way it may be worth a proper entity, but
it works now and needs nothing from me.

Thank you for chasing this properly rather than leaving it — between you and
@AdrianCzebiniak, a feature request that sat for two years is closed and the
people it does serve have it.

**Comment by Hukuma1** [comment]
Did some digging, it must be some other issue but hopefully the other people in this thread can comment if your code worked on their end.

Found this thread: https://community.home-assistant.io/t/privacy-mode-on-amcrest-ip-cameras-via-a-switch/287543

Which gives the exact commands to execute:

`curl -g --digest -u <user>:<password> "http://<camera-ip>/cgi-bin/configManager.cgi?action=setConfig&LeLensMask[0].Enable=true"`

That command worked for me to turn it on. False to turn it off.

So it looks like "LeLensMask" is standard privacy mode for Amcrest and it doesn't matter if the cover mechanism is physical or not.

The workaround you mentioned with different preset PTZ is what I've been doing for the last 2 years, however the big limitation there is it doesn't cut the feed/mic. But I guess ultimately I can just wire up a switch manually in HA like that thread suggested. They ended up upstreaming that code into the official Amcrest integration code. That's the code I was hoping to originally merge into this repo.

**Comment by zios0007** [comment]
That is the answer, and it is the good branch of the two I gave you. Your camera
**does** have `LeLensMask` — it is reachable over CGI and simply refused over
RPC2, which is why the error said `Authority:check failure` and why granting
admin changed nothing. The permissions were never the problem.

So this is a defect in #682, not a limitation of your hardware, and you found it.

**#688 fixes it**: CGI first, RPC2 as the fallback. Reasons, in order of how much
they matter:

1. Your camera works over CGI. That is the whole point.
2. CGI is what the Amcrest integration uses for this, which is the code you were
   asking to have ported in the first place — so the route you wanted is now the
   primary one.
3. It costs no login of its own. The RPC2 path logs in and out around every
   single call, which I had already flagged for review when I rebased #682; this
   removes it for every camera CGI can read.

RPC2 stays as the fallback so @akominch's DH-SD-H4C-0400B, and anything else that
answers only there, keeps working.

**Correcting myself twice over.** I told you your camera probably had no lens
mask hardware and pointed you at a PTZ workaround. Wrong on the first count —
and on the second, you had already been doing exactly that for two years and told
me its real limitation, that it does not cut the feed. I should have taken the
"my camera has no cover but the app still does *something*" report at face value
and asked what the app was writing, rather than concluding the feature was absent.

**One thing worth knowing about what you will get.** The switch reads and writes
`LeLensMask[0].Enable`, the same key your curl proved. On your camera that
appears to drive the position-and-cut-feed behaviour rather than a physical
cover, which is fine — the integration does not care what the camera does with
it, only that the table is there and the state reads back.

Once #688 is merged and released, a **Privacy Mode** switch should appear for
that camera without you doing anything. If it does not, the line to look for at
debug is `Device supports privacy mode=`, and I would want to know.

**Comment by Hukuma1** [comment]
Fantastic. Will test it out once it's out. Thank you!

I hope the others can chime in as well if it works on their end. I hope we didn't fix my issue to break theirs. ;)

**Comment by zios0007** [comment]
Fair worry, and worth answering rather than hand-waving.

**It cannot regress a confirmed-working setup, because there isn't one.** Privacy
mode has never been verified on hardware by anybody — #516 was written in October
2025 and never tested by its author or by @rroller, and when I rebased it as #682
I said plainly that I have no camera with `LeLensMask`. You are the first person
to put a real camera against it, and it did not work. So the RPC2 path has never
been shown to work on any device either.

**And the fallback is only reached when CGI cannot answer.** A camera whose
`configManager.cgi?action=getConfig&name=LeLensMask` returns nothing, or errors,
drops through to exactly the RPC2 code that is in 0.10.2 today, unchanged. So the
worst case for @akominch and @beckynet is that they get what they have now.

The realistic case is better than that: if their cameras answer CGI — and it is a
plain config table, so they very likely do — they get a path that needs no login
per call, which is strictly less work for the camera than what ships today.

The thing that would genuinely help is what you are already suggesting: someone
with a **physical** lens cover confirming it once #688 is released. @akominch
opened this issue in 2024 with a DH-SD-H4C-0400B whose log shows `leLensMask`, and
@beckynet has a DH-H4C. Between your camera and one of theirs we would have both
kinds of hardware covered, which is more than this feature has ever had.

**Comment by Hukuma1** [comment]
Can confirm it works on my end! Automatically added the switch to the device page and executes as expected. Thank you!
